test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala

```diff
@@ -195,4 +209,35 @@ class ArrayTest {
 
     assertThrows(classOf[NegativeArraySizeException], testCreateNegativeSizeArray())
   }
+
+  @Test def genericArrayNullsShortCircuited_Issue4755(): Unit = {
+    // Tests for the intrinsics for ScalaRunTime.array_{apply,update,select}.
+
+    @inline def testGeneric[T](array: Array[T], value: T): Unit = {
+      assertThrows(classOf[IllegalStateException], array(throwIllegalStateAsInt()))
+      assertThrows(classOf[IllegalStateException], array(throwIllegalStateAsIntInline()))
+
+      assertThrows(classOf[IllegalStateException], array(throwIllegalStateAsInt()) = value)
+      assertThrows(classOf[IllegalStateException], array(throwIllegalStateAsIntInline()) = value)
+
+      assertThrows(classOf[IllegalStateException], array(1) = (throw new IllegalStateException()))
+    }
+
+    @noinline def testNoInline[T](array: Array[T], value: T): Unit = {
+      testGeneric(array, value)
+    }
+
+    @inline def test[T](array: Array[T], value: T): Unit = {
+      testNoInline(array, value)
+      testGeneric(array, value)
+    }
+
+    val nullArrayRef = nullOf[Array[AnyRef]]
```

Comment:
Nit: Inline this val? (also below).

Reply:
The `val`s are actually necessary for the test not to be moot. I added a comment about that.